flowmachine/flowmachine/core/subscriber_subsetter.py

```diff
@@ -0,0 +1,249 @@
+        # called from Query.dependency_graph(). However, since this
+        # class doesn't really represent a full flowmachine.Query
+        # (and this inheritance will be removed in the long run)
+        # we can't return meaningful values here, so we just return
+        # a dictionary with the correct keys but no actual values.
+        attrs = {}
+        attrs["name"] = self.__class__.__name__
+        attrs["stored"] = "N/A"
+        attrs["cost"] = "N/A"
+        attrs["runtime"] = "N/A"
+        return attrs
+
+
+class SubscriberSubsetterForAllSubscribers(SubscriberSubsetterBase):
+    """
+    Represents the subset of all subscribers - i.e., no subsetting at all.
+    In other words this is the "null object" for the subsetting logic, which
+    represents the case of "no work needed".
+
+    The reason this exists is so that external code does not need to know
+    anything about the subsetting logic or implementation and can use any
+    of these classes completely interchangeably, no matter what kind
+    of subset is required or whether any subsetting is needed at all.
+
+    If we didn't have this class then external code would need to make
+    a case distinction to check if subsetting is needed, which leads
+    to coupling between unrelated parts of the code base, unnecessary
+    complexity and makes testing of all possible cases much difficult.
+    """
+
+    is_proper_subset = False
+
+    def _make_query(self):
+        # Return a dummy string representing this subset. This is only needed
+        # because SubscriberSubsetterBase currently inherits from Query, but will
+        # eventually be removed.
+        return "<SubscriberSubsetterForAllSubscribers>"
+
+    def apply_subset_if_needed(self, sql, *, subscriber_identifier=None):
+        """
+        Return the input query unchanged, since no subsetting is applied.
+
+        Parameters
+        ----------
+        sql : sqlalchemy.sql.ClauseElement
+            The SQL query to which the subset should be applied.
+
+        subscriber_identifier : str
+            This argument is ignored for subsets of type 'AllSubscribers'.
+        """
+        return sql
+
+
+class SubscriberSubsetterForFlowmachineQuery(SubscriberSubsetterBase):
+    """
+    Represents a subset given by a flowmachine query.
+    """
+
+    is_proper_subset = True
+
+    def __init__(self, flowmachine_query):
+        """
+        Parameters
+        ----------
+        flowmachine_query : flowmachine.Query
+            The flowmachine query to be used for subsetting. The only requirement
+            on it is that the result has a column called "subscriber" (it is fine
+            for other columns to be present, too).
+        """
+        assert isinstance(flowmachine_query, Query)
+
+        self._verify_that_subscriber_column_is_present(flowmachine_query)
+        self.flowmachine_query = flowmachine_query
+
+    def _verify_that_subscriber_column_is_present(self, flowmachine_query):
```

Comment:
Not sure if this might be a replication of [this](https://github.com/Flowminder/FlowKit/blob/1334dcec5a237b45c5078790ea3a5fae0b67eab8/flowmachine/flowmachine/utils/utils.py#L245) which will be soon merged.

Reply:
Yes, I think it is. Good spot. I'll change it once #320 has been merged.